### PR TITLE
Improved keras imports

### LINKED ISSUE #24437: TFPreTrainedModel.build breaks pytorch PreTrainedModel.from_pretrained(from_tf=True)
### System Info

- `transformers` version: 4.30.2
- Platform: macOS-13.4-x86_64-i386-64bit
- Python version: 3.9.16
- Huggingface_hub version: 0.14.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.1 (False)
- Tensorflow version (GPU?): 2.10.1 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: NO
- Using distributed or parallel set-up in script?: NO

### Who can help?

@Rocketknight1 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The `build()` overridden method introduced recently breaks `PreTrainedModel.from_pretrained(from_tf=True)`. This can be reproduced by the following line of codes:

```python
model = RobertaModelForSequenceClassification.from_pretrained(path, from_tf=True)
```
console prints out:
```
All TF 2.0 model weights were used when initializing RobertaForSequenceClassification.

Some weights of RobertaForSequenceClassification were not initialized from the TF 2.0 model and are newly initialized: ['roberta.embeddings.word_embeddings.weight', 'roberta.embeddings.position_embeddings.weight', 'roberta.embeddings.token_type_embeddings.weight', ... all 6 layers... 
```

After some digging, I believe the recursive call to `__call__` in the `build()` causes the names of all TF weights to be prefixed twice with the model's keras name (instead of once,) e.g. `tf_roberta_for_sequence_classification/tf_roberta_for_sequence_classification/...`. The `from_pretrained(from_tf=True)` works according to the following steps:

1. create a PT model
2. create a TF model
3. build the TF model by calling `tf_model(tf_mode.dummy_inputs, training=False)`
    1. calling `tf_model._call_` 
    2. enter name scope of the model, e.g. `tf_roberta_for_sequence_classification`
    3. figure it has not been built because `built=False` 
    4. call `tf_model.build` 
    5. the overridden `TFPreTrainedModel.build` then set `built=True` 
    6. calls `self._call__` (i.e. `tf_model.__call__`) again
    7. **enter name scope of the model again**, e.g. e.g. => `tf_roberta_for_sequence_classification/tf_roberta_for_sequence_classification`
    8. proceed to call the `tf_model.call`
    9. call the layer’s `_call_` 
    10. add variables with name e.g. `tf_roberta_for_sequence_classification/tf_roberta_for_sequence_classification/roberta/...`
4. load TF weights to TF models
5. map TF weight names to PT weight names by removing the **first** prefix of the TF variable names and copy over the weights. => fail to map TF names to PT names due to the double prefix e.g. => `tf_roberta_for_sequence_classification/tf_roberta_for_sequence_classification`

The hacky workaround is to override the `build` ourself to not call the `__call__`, e.g.:

```python
class _FixedTFRobertaForSequenceClassification(transformers.TFRobertaForSequenceClassification):

    def build(self, input_shape=None):
        self.built = True

    if id(transformers.TFRobertaForSequenceClassification) != id(_FixedTFRobertaForSequenceClassification):
        print('fixing TFRobertaForSequenceClassification to', _FixedTFRobertaForSequenceClassification.__name__)
        transformers.TFRobertaForSequenceClassification = _FixedTFRobertaForSequenceClassification
    model = RobertaForSequenceClassification.from_pretrained(language_model_path, from_tf=True)

All TF 2.0 model weights were used when initializing RobertaForSequenceClassification.

All the weights of RobertaForSequenceClassification were initialized from the TF 2.0 model
```

The double naming can also be seen by just creating the submodel:
```python
config = RobertaConfig.from_pretrained(language_model_path)
tf_model = TFRobertaForSequenceClassification(config)
print(tf_model._name)
tf_model.weights[10]
# => notice the double prefix in the weight variable names
```

There should be better fix.

CC @Rocketknight1 

### Expected behavior

1. TF submodel variables/weights names should not be double-prefixed
2. `from_pretrained(from_tf=True)` should works.

### PR DESCRIPTION
**Pull request by Rocketknight1**
A sneaky bug was hiding in our Keras imports, where an import for `call_context` appeared to succeed on some TF versions, but actually got an older, unusable version of the function. This caused `build()` to behave improperly in some cases.

I went on a quest to fix this, and generally clean up our version-specific imports for TensorFlow to stop things like this from happening in future. I also bumped the minimum version for TF to 2.6 (2.6 should be 2 years old by the time of our next release), and eliminated the version cap in our dependency table because TF 2.13 should also be fully supported now.

This involved a partial rewrite of some code, where we checked for `KerasTensor` in a lot of places. However, this is an internal class for Keras and they keep moving it around, so trying to import it feels like a bad idea. Instead, I'm relying on `tf.is_tensor()`, which returns `True` for anything tensor-y, including symbolic tensors and Keras `Input` placeholders.

Fixes #24437

### CODE REVIEW

**Comment by ydshieh** [comment]
> Keras and they keep moving it around

It's tensor-flow ... that's probably the reason.

**Comment by Rocketknight1** [comment]
I tested this with a spread of TF versions going back to 2.6. Even at 2.6, it was pretty hard to get an environment that worked with modern `transformers` - old TensorFlow keeps trying to use NumPy features that have been deprecated and deleted, but our modern libraries need a minimum version of NumPy to run at all, so there's actually only a very narrow window of NumPy versions that can even run both at once! I think going back to 2.5 or earlier would be very difficult, so I'm pretty comfortable with bumping our minimum version at this point.

In all versions I tested with this patch, our test suite runs well and the issue identified in #24437 is fixed.

File: setup.py [hunk 1]
```diff
@@ -168,9 +168,9 @@
     "sudachipy>=0.6.6",
     "sudachidict_core>=20220729",
     # TensorFlow pin. When changing this value, update examples/tensorflow/_tests_requirements.txt accordingly
-    "tensorflow-cpu>=2.4,<2.13",
-    "tensorflow>=2.4,<2.13",
-    "tensorflow-text<2.13",
+    "tensorflow-cpu>=2.6",
```
Comment (sgugger) [inline setup.py:171]: No let's leave a maximum pin. I do not want any surprises.

**Review by sgugger — COMMENTED** [review]
Thanks for the cleanup. I will agree to not put a maximum pin on TensorFlow when they stop making breaking changes in minor releases (so never 😅 )

File: setup.py (hunk 1 above)
Comment (Rocketknight1) [reply setup.py:171]: Deal, but I'm bumping the cap to `<2.14`, since 2.13 works fine!

Comment (sgugger) [reply setup.py]: Sure!

**Review by sgugger — APPROVED** [review]

**Pull request merged**